### ISSUE TITLE
Certification failed due to offset report not correct when playing music

### ISSUE THREAD

**Issue by subjectxbj** [alexa/avs-device-sdk#973]
**IMPORTANT**: Before you create an issue, please take a look at our [Issue Reporting Guide](https://github.com/alexa/avs-device-sdk/wiki/Issue-Reporting-Guide).

## Briefly summarize your issue:
AVS certification team reported some blocker bugs on our unit. Some are related with the offiset report when playing music.

Issue 1: "progressReportDelayElapsed" is sent before "playbackStarted". 

I checked the code, seems the timer of "progressReportDelayElapsed" is started right after receiving the "Play" directive. But actually the alexa client will take some seconds to download the audio stream and then start play and send "playbackStarted" once the audio is actually played.  So if the value of "progressReportDelayInMilliseconds" is small, then it is very likely that the "progressReportDelayElapsed"is sent before "playbackStarted". 

So is this really a bug?

Also seems the offset in playbackFinished event is not correct.
{noformat}
 Actual logs : 

17:29:12.542 play the song the devil in i by slipknot from my music library 
{color:#d04437}17:29:25.459 progressReportDelayElapsed 1764 - sent before playbackStarted {color} 
17:29:26.685 playbackStarted 0 
17:29:51.101 stop 
17:29:53.371 playbackStopped 25774 
17:30:14.128 alexa resume 
17:30:19.161 playbackStarted 25774 
{color:#d04437}17:35:38.331 playbackNearlyFinished 24250 - wrong offset{color} 
{color:#d04437}17:35:38.636 playbackFinished 24250 - wrong offset{color} 

Expected logs :  

17:29:12.542 play the song the devil in i by slipknot from my music library 
17:29:26.685 playbackStarted 0 
17:29:51.101 stop 
17:29:53.371 playbackStopped 25774 
17:30:14.128 alexa resume 
17:30:19.161 playbackStarted 25774 
{color:#d04437}17:35:38.331 playbackNearlyFinished 344774 - correct offset{color} 
{color:#d04437}17:35:38.636 playbackFinished 344774 - correct offset{color} 

{noformat}

I checked the AVSonline document, it mentioned :+1: 
https://developer.amazon.com/zh/docs/alexa-voice-service/audioplayer.html#progressreportintervalelapsed
```
 For example, if the Play directive contains progressReportDelayInMilliseconds with a value of 20000, the ProgressReportDelayElapsed event must be sent 20,000 milliseconds from the start of the track. However, 
```
That means the ProgressReportDelayElapsed event should be sent start from the real audio play. In that case, I think this is e SDK bug.
### What version of the AVS Device SDK are you using?
      <x.y.z>

### Tell us what hardware you're using:
- [ ] Desktop / Laptop
- [ ] Raspberry Pi
- [ ] Other - tell us more:

### Tell us about your OS (Type & version):
- [ ] Linux
- [ ] MacOS 
- [ ] Raspbian Stretch
- [ ] Raspbian Jessy
- [ ] Other - tell us more:

**Comment by subjectxbj** [comment]
I put the client side log:
```
2018-09-20 09:30:23.243 [ 13] 9 AudioPlayer:PLAY:payload={"audioItem"\:{"audioItemId"\:"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#1ea5a934-c149-4bfb-aa50-817c5bf0156b\:1"\,"stream"\:{"offsetInMilliseconds"\:0\,"expiryTime"\:"2018-09-20T09\:31\:18+0000"\,"progressReport"\:{"progressReportDelayInMilliseconds"\:1000}\,"url"\:"https\://d29r7idq0wxsiz.cloudfront.net/DigitalMusicDeliveryService/HPS.m3u8?m\=m&dmid\=250453756&c\=cf&f\=ts&t\=10&bl\=256k&s\=true&e1\=1537437600000&e2\=1537437600000&v\=V2&n\=PSLZCFV4MKJSC&h\=6064107004c71363886b2d8007dc7b65fb8a12e1754a33f04e52f64bd4761c3f"\,"token"\:"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#1ea5a934-c149-4bfb-aa50-817c5bf0156b\:1"}}\,"playBehavior"\:"REPLACE_ALL"}
2018-09-20 09:30:23.245 [ 13] E JsonUtils:findNodeFailed:reason=missingDirectChild,child=streamFormat
2018-09-20 09:30:23.246 [ 13] E JsonUtils:findNodeFailed:reason=missingDirectChild,child=progressReportIntervalInMilliseconds
2018-09-20 09:30:23.248 [ 13] E JsonUtils:findNodeFailed:reason=missingDirectChild,child=expectedPreviousToken
...............
2018-09-20 09:30:24.452 [  c] 1 AudioPlayer:sendEventWithTokenAndOffset:eventName=ProgressReportDelayElapsed
2018-09-20 09:30:24.454 [  c] 9 MediaPlayer:getOffsetCalled
2018-09-20 09:30:24.455 [  2] 9 AttachmentReaderSource:read:size=0,status=1
2018-09-20 09:30:24.456 [  2] 0 MediaPlayer:handleGetOffsetCalled:idPassed=6,currentId=6
2018-09-20 09:30:24.457 [  2] I MediaPlayer:handleGetOffset:reason=getElementGetStateUnsuccessful,stateChangeReturn=ASYNC
2018-09-20 09:30:24.459 [  c] 0 EventBuilder:buildJsonEventString:messageId=476bd575-4968-4b84-9cdb-4de7c85ff75e,namespace=AudioPlayer,name=ProgressReportDelayElapsed
2018-09-20 09:30:24.460 [  c] 9 HTTP2Transport:enqueueRequest:jsonContent={"event"\:{"header"\:{"namespace"\:"AudioPlayer"\,"name"\:"ProgressReportDelayElapsed"\,"messageId"\:"476bd575-4968-4b84-9cdb-4de7c85ff75e"}\,"payload"\:{"token"\:"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#1ea5a934-c149-4bfb-aa50-817c5bf0156b\:1"\,"offsetInMilliseconds"\:1432827}}}
.................
2018-09-20 09:30:30.179 [  c] 1 AudioPlayer:notifyObserver:playerActivity=PLAYING
2018-09-20 09:30:30.182 [  c] 1 AudioPlayer:sendEventWithTokenAndOffset:eventName=PlaybackStarted
2018-09-20 09:30:30.183 [  c] 0 EventBuilder:buildJsonEventString:messageId=2808b4ca-8d43-4b30-8933-b32fd6c0e03f,namespace=AudioPlayer,name=PlaybackStarted
2018-09-20 09:30:30.179 [  2] 9 AttachmentReaderSource:read:size=4096,status=0
2018-09-20 09:30:30.184 [  c] 9 HTTP2Transport:enqueueRequest:jsonContent={"event"\:{"header"\:{"namespace"\:"AudioPlayer"\,"name"\:"PlaybackStarted"\,"messageId"\:"2808b4ca-8d43-4b30-8933-b32fd6c0e03f"}\,"payload"\:{"token"\:"amzn1.as-ct.v1.Dee-Domain-Music#ACRI#url#ACRI#1ea5a934-c149-4bfb-aa50-817c5bf0156b\:1"\,"offsetInMilliseconds"\:0}}}

```

**Comment by scotthea-amazon** [comment]
Hello subjectxbj,

Thank you for bringing this to our attention!  Can you tell us what version your are experiencing this issue with?

Thank you,
-SWH

**Comment by scotthea-amazon** [comment]
Hello again subjectxbj,

I was able to reproduce this with v1.9.0.  We will continue to investigate.

Thank you again,
-SWH

**Comment by subjectxbj** [comment]
@scotthea-amazon,
Thanks for your feedback. I think this issue exist in all releases. Because this part of code seems not changed acrosss all these releases.

Thanks

**Comment by visayamv** [comment]
Hi @subjectxbj , can you tell me which release version did you use for this case? In v1.90, there was a fix for the incorrect playbackfinished offset.

**Comment by visayamv** [comment]
The out of order issue could be different

**Comment by subjectxbj** [comment]
> you tell me which release version did you use for this case? I

Hi @visayamv , I'm using 1.9.0. So it seems this issue still exist on 1.9.0.

**Comment by visayamv** [comment]
@subjectxbj , 
When you say:
`When I pause the music and resume the music, it can get and report a reasonable offset`
Do you mean if you just play the song without doing anything the offset in the playbackfinished event is incorrect, but if you pause and then resume in the middle of playing the music the offset in the playbackfinished event is correct?

**Comment by visayamv** [comment]
Hi @subjectxbj , based on your log you have received an error when the playbackfinished event occurred and the offset is being queried from the gstreamer pipeline:

`................Play finished...................................
2018-09-20 09:36:12.923 [  2] 0 AudioPlayer:onPlaybackFinished:id=6
2018-09-20 09:36:12.924 [  2] 9 MediaPlayer:tearDownTransientPipelineElements
2018-09-20 09:36:12.924 [  c] 1 AudioPlayer:executeOnPlaybackFinished:id=6
2018-09-20 09:36:12.925 [  2] E MediaPlayer:saveOffsetBeforeTeardown - gst_element_query_position failed
`
I have tried to reproduce this case, but I was unable to. My step is just by asking 'Alexa play the devil and I by slipknot from Amazon Music' and wait for the music to end.

I was hoping if you can provide additional logging for this as I try to investigate the state of the gstreamer pipeline when this issue happens.

Can you please open the Gstreamer verbose logging and also submit ALL the debug logs generated by SampleApp?

You can do the following:

export GST_DEBUG=6

#On your sample app build directory
./SampleApp -C <your AlexaConfig Json path> -L debug9 2>&1 | tee debug.log

Please reproduce the failure case and upload the debug.log file here.

Thanks!

**Comment by subjectxbj** [comment]
> @subjectxbj ,
> When you say:
> `When I pause the music and resume the music, it can get and report a reasonable offset`
> Do you mean if you just play the song without doing anything the offset in the playbackfinished event is incorrect, but if you pause and then resume in the middle of playing the music the offset in the playbackfinished event is correct?

Yes, that is right

**Comment by subjectxbj** [comment]
> Hi @subjectxbj , based on your log you have received an error when the playbackfinished event occurred and the offset is being queried from the gstreamer pipeline:
> 
> `................Play finished................................... 2018-09-20 09:36:12.923 [ 2] 0 AudioPlayer:onPlaybackFinished:id=6 2018-09-20 09:36:12.924 [ 2] 9 MediaPlayer:tearDownTransientPipelineElements 2018-09-20 09:36:12.924 [ c] 1 AudioPlayer:executeOnPlaybackFinished:id=6 2018-09-20 09:36:12.925 [ 2] E MediaPlayer:saveOffsetBeforeTeardown - gst_element_query_position failed`
> I have tried to reproduce this case, but I was unable to. My step is just by asking 'Alexa play the devil and I by slipknot from Amazon Music' and wait for the music to end.
> 
> I was hoping if you can provide additional logging for this as I try to investigate the state of the gstreamer pipeline when this issue happens.
> 
> Can you please open the Gstreamer verbose logging and also submit ALL the debug logs generated by SampleApp?
> 
> You can do the following:
> 
> export GST_DEBUG=6
> 
> #On your sample app build directory
> ./SampleApp -C "your AlexaConfig Json path" -L debug9 2>&1 | tee debug.log
> 
> Please reproduce the failure case and upload the debug.log file here.
> 
> Thanks!

Hi @visayamv ,
I will get back to office next week to try a new build. But for the error log "saveOffsetBeforeTeardown - gst_element_query_position failed". If you take it as the key and search in the all the issues. You will find it is very commonly exist in the logs that are provided by a lot of people. I also noticed this error, but not sure if that is the root cause of this problem.

**Comment by visayamv** [comment]
Hi @subjectxbj , that error can appear when the music is paused or has not started yet. It should not fail when the music finishes. I have compared my logs to yours and also noticed this error. The gst_element_query_position  failure on saveOffserBeforeTeardown does not appear in my log when the song finishes so it was able to get the offset value properly. But when the song starts or is paused that error appears

**Comment by celinval** [comment]
Hi @subjectxbj, have you had the chance to try the new build yet?

Thanks!

**Comment by subjectxbj** [comment]
Hi @celinval ,
I tested again, and find there is no problem anymore when playback finished. Seems this issue does not happen everytime.
```
2018-10-23 08:26:16.804 [  4] 9 MediaPlayer:messageReceived:type=eos,source=audio-pipeline
2018-10-23 08:26:16.806 [  4] 0 MediaPlayer:callingOnPlaybackFinished:currentId=7
2018-10-23 08:26:16.807 [  4] 0 AudioPlayer:onPlaybackFinished:id=7
2018-10-23 08:26:16.808 [  4] 9 MediaPlayer:tearDownTransientPipelineElements
2018-10-23 08:26:16.808 [ 15] 1 AudioPlayer:executeOnPlaybackFinished:id=7
2018-10-23 08:26:16.814 [  4] 0 MediaPlayer:saveOffsetBeforeTeardown:offset=343589
2018-10-23 08:26:16.815 [ 15] 0 AudioPlayer:changeActivity:from=PLAYING,to=FINISHED
2018-10-23 08:26:16.817 [ 15] 0 AudioPlayer:executeProvideState:sendToken=false,stateRequestToken=0
2018-10-23 08:26:16.819 [ 15] 9 MediaPlayer:getOffsetCalled

```

**Comment by celinval** [comment]
Happy to hear @subjectxbj! I'm going to close this issue but feel free to reopen it if you find the same problem again.

**Comment by visayamv** [comment]
@subjectxbj , the out-of-order and incorrect offsets in progressReport* events have been fixed in 1.10. So please upgrade to 1.10.
